src/abi_impl.rs

```diff
@@ -417,6 +417,37 @@ pub(crate) fn assembly_script_get_time(env: &Env) -> ABIResult<i64> {
     }
 }
 
+/// sends an async call
+pub(crate) fn assembly_script_send_message(
+    env: &Env,
+    target_address: i32,
+    target_handler: i32,
+    validity_start_period: i64,
+    validity_start_thread: i64,
+    validity_end_period: i64,
+    validity_end_thread: i64,
+    max_gas: i64,
+    gas_price: i32,
+    coins: i32,
+    payload: i32,
+) -> ABIResult<()> {
+    sub_remaining_gas(env, settings::metering_send_message())?;
+    let memory = get_memory!(env);
+    match env.interface.send_message(
+        &get_string(memory, target_address)?,
+        &get_string(memory, target_handler)?,
+        (validity_start_period as u64, validity_start_thread as u64),
```

Comment:
all those i64 -> u64 conversions will panic and shut the whole program if an attacker calls the function with negative values. Use try_from.

Reply:
Note that the "thread" values should convert to u8 and not u64